File: soc-server/internal/agent_grpc/handler.go
// ==============================================================================
// Package agent_grpc - gRPC Service Handler
// File: internal/agent_grpc/handler.go
// Mô tả: Implement AgentService gRPC với 2 RPC methods:
//
//   1. StreamEvents (Bidirectional Streaming):
//      - Goroutine 1 (Receiver): Nhận stream Sysmon/Event Log từ Agent
//        → Parse JSON → Gọi RuleEngine.EvaluateLog() → Tạo Alert → Broadcast
//      - Goroutine 2 (Sender): Lắng nghe command channel
//        → Đẩy CommandResponse (KILL_PROCESS, BLOCK_IP) xuống Agent
//
//   2. Heartbeat (Unary):
//      - Cập nhật last_heartbeat_at trong PostgreSQL
//      - Set agent status = "online" trong Redis (với TTL)
// ==============================================================================

package agent_grpc

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"time"

	pb "soc-server/internal/generated/proto"
	"soc-server/internal/models"
	"soc-server/internal/rules"
	"soc-server/pkg/database"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// AlertCallback - Function type để gọi callback khi có Alert mới
// Cho phép handler thông báo cho các module khác (WebSocket broadcast, SOAR dispatch)
type AlertCallback func(alert *models.Alert)

// AgentServiceHandler - Implement AgentServiceServer interface từ protobuf
type AgentServiceHandler struct {
	pb.UnimplementedAgentServiceServer // Embed default implementation

	db          *gorm.DB              // PostgreSQL connection
	redis       *database.RedisClient // Redis client
	ruleEngine  *rules.RuleEngine     // Detection Rule Engine
	connManager *ConnectionManager    // Connection Manager
	onNewAlert  AlertCallback         // Callback khi có Alert mới
}

// NewAgentServiceHandler - Khởi tạo handler với tất cả dependencies
func NewAgentServiceHandler(
	db *gorm.DB,
	redis *database.RedisClient,
	ruleEngine *rules.RuleEngine,
	connManager *ConnectionManager,
	onNewAlert AlertCallback,
) *AgentServiceHandler {
	return &AgentServiceHandler{
		db:          db,
		redis:       redis,
		ruleEngine:  ruleEngine,
		connManager: connManager,
		onNewAlert:  onNewAlert,
	}
}

// ==============================================================================
// StreamEvents - Bidirectional Streaming RPC
//
// Luồng xử lý:
//
//	Agent ──stream EventRequest──→ Server (Goroutine Receiver)
//	                                 ↓ EvaluateLog()
//	                                 ↓ Match? → Create Alert → Broadcast
//	Agent ←──stream CommandResponse── Server (Goroutine Sender)
//	                                 ↑ Lắng nghe CmdChan
//
// ==============================================================================
func (h *AgentServiceHandler) StreamEvents(stream pb.AgentService_StreamEventsServer) error {
	// Tạo connection mới cho Agent
	conn := &ActiveConnection{
		Stream:  stream,
		CmdChan: make(chan *pb.CommandResponse, 100), // Buffer 100 commands
		Done:    make(chan struct{}),
	}

	// Biến flag để biết đã register chưa
	registered := false
	var agentID string

	// ===== Goroutine 1: Receiver - Nhận events từ Agent =====
	// Goroutine này chạy trong main goroutine của StreamEvents
	// Khi stream đóng (Agent disconnect), hàm sẽ return
	defer func() {
		if registered {
			h.connManager.Unregister(agentID)
			// Cập nhật trạng thái Agent offline trong Redis
			ctx := context.Background()
			_ = h.redis.SetAgentOffline(ctx, agentID)
			// Cập nhật trong database
			h.db.Model(&models.Agent{}).Where("id = ?", agentID).Update("status", "offline")
			log.Printf("[STREAM] 🔴 Agent '%s' đã ngắt kết nối stream", agentID)
		}
	}()

	// ===== Goroutine 2: Sender - Đẩy commands xuống Agent =====
	// Goroutine riêng lắng nghe channel CmdChan và gửi qua stream
	go func() {
		for {
			select {
			case cmd := <-conn.CmdChan:
				// Gửi command xuống Agent qua gRPC stream
				if err := stream.Send(cmd); err != nil {
					log.Printf("[STREAM] ❌ Lỗi gửi command xuống Agent '%s': %v", agentID, err)
					return
				}
				log.Printf("[STREAM] 📤 Đã gửi command '%s' xuống Agent '%s'",
					cmd.CommandType.String(), agentID)
			case <-conn.Done:
				// Kết nối đã đóng, dừng goroutine sender
				return
			}
		}
	}()

	// ===== Main Loop: Nhận events từ Agent stream =====
	for {
		// Nhận EventRequest từ Agent (blocking)
		event, err := stream.Recv()
		if err == io.EOF {
			log.Printf("[STREAM] Agent '%s' đã đóng stream (EOF)", agentID)
			return nil
		}
		if err != nil {
			log.Printf("[STREAM] ❌ Lỗi nhận event từ Agent '%s': %v", agentID, err)
			return err
		}

		// Lần nhận đầu tiên: Register Agent connection
		if !registered {
			agentID = event.AgentId
			conn.AgentID = agentID
			conn.Hostname = event.Hostname
			conn.IPAddress = event.IpAddress

			// Đăng ký connection vào Connection Manager
			h.connManager.Register(conn)
			registered = true

			// Cập nhật trạng thái Agent online
			ctx := context.Background()
			_ = h.redis.SetAgentOnline(ctx, agentID)
			var agent models.Agent
			result := h.db.Where("id = ?", agentID).First(&agent)
			switch result.Error {
			case nil:
				h.db.Model(&agent).Updates(map[string]interface{}{
					"status":     "online",
					"ip_address": event.IpAddress,
					"hostname":   event.Hostname,
				})
			case gorm.ErrRecordNotFound:
				h.db.Create(&models.Agent{
					ID:        agentID,
					Hostname:  event.Hostname,
					IPAddress: event.IpAddress,
					Status:    "online",
				})
			default:
				log.Printf("[STREAM] ⚠️ Không thể lưu Agent '%s': %v", agentID, result.Error)
			}

			log.Printf("[STREAM] 🟢 Agent '%s' (%s) đã bắt đầu stream events", agentID, event.Hostname)
		}

		// ===== Xử lý Event: Parse JSON → Evaluate Rule =====
		h.processEvent(event, agentID)
	}
}

// processEvent - Xử lý 1 event nhận từ Agent
// Parse raw_payload JSON → Gọi RuleEngine.EvaluateLog() → Tạo Alert nếu match
func (h *AgentServiceHandler) processEvent(event *pb.EventRequest, agentID string) {
	// Parse raw_payload (JSON string) thành map
	var rawLog map[string]interface{}
	if err := json.Unmarshal([]byte(event.RawPayload), &rawLog); err != nil {
		log.Printf("[STREAM] ⚠️ Lỗi parse raw_payload JSON từ Agent '%s': %v", agentID, err)
		// Nếu không parse được JSON, tạo map thủ công từ fields có sẵn
		rawLog = map[string]interface{}{
			"raw_data": event.RawPayload,
		}
	}

	// Bổ sung metadata vào rawLog để Rule Engine có thêm context
	rawLog["event_type"] = event.EventType
	rawLog["hostname"] = event.Hostname
	rawLog["ip_address"] = event.IpAddress
	rawLog["agent_id"] = agentID

	// Metric là telemetry phục vụ biểu đồ/trạng thái Agent, không phải security alert.
	if event.EventType == "system_metric" {
		return
	}
	if event.EventType == "file_integrity" {
		log.Printf("[STREAM] 📁 Nhận FIM event từ Agent '%s': %s", agentID, event.RawPayload)
	}

	// ===== Gọi Rule Engine đánh giá event =====
	alert, matched := h.ruleEngine.EvaluateLog(rawLog, agentID)
	if !matched {
		// Event không nguy hiểm vẫn được lưu để analyst theo dõi trên dashboard.
		rawPayloadJSON, _ := json.Marshal(rawLog)
		alert = &models.Alert{
			ID:          uuid.New().String(),
			AgentID:     agentID,
			RuleID:      "EVENT-OBSERVED",
			EventType:   models.AlertEventType(event.EventType),
			Severity:    models.SeverityLow,
			RawPayload:  string(rawPayloadJSON),
			Status:      models.AlertStatusNew,
			Title:       fmt.Sprintf("[LOW] Event observed: %s", event.EventType),
			Description: "Sự kiện được ghi nhận từ Agent nhưng không khớp detection rule nguy hiểm.",
			CreatedAt:   time.Now(),
		}
	}

	// ===== Rule matched! Lưu Alert vào database =====
	if err := h.db.Create(alert).Error; err != nil {
		log.Printf("[STREAM] ❌ Lỗi lưu Alert vào database: %v", err)
		return
	}

	log.Printf("[STREAM] 🚨 ALERT CREATED: [%s] %s (Agent: %s, Rule: %s)",
		alert.Severity, alert.Title, agentID, alert.RuleID)

	// ===== Ghi Audit Log =====
	payloadSummary, _ := json.Marshal(map[string]string{
		"event_type": string(alert.EventType),
		"rule_id":    alert.RuleID,
		"severity":   string(alert.Severity),
	})
	auditLog := &models.AuditLog{
		ID:             uuid.New().String(),
		EventType:      "alert_created",
		Source:         models.AuditSourceRuleBased,
		ActionTaken:    models.ActionLogOnly,
		Confidence:     1.0,
		Actor:          "rule_engine",
		PayloadSummary: string(payloadSummary),
		RelatedAlertID: alert.ID,
		RelatedAgentID: agentID,
		CreatedAt:      time.Now(),
	}
	h.db.Create(auditLog)

	// ===== Gọi callback thông báo Alert mới =====
	// Callback sẽ: 1) Broadcast WebSocket, 2) Dispatch n8n
	if h.onNewAlert != nil {
		go h.onNewAlert(alert) // Non-blocking: chạy trong goroutine riêng
	}
}

// ==============================================================================
// Heartbeat - Unary RPC
// Agent gửi heartbeat định kỳ (mỗi 30 giây) để báo cáo trạng thái
// ==============================================================================
func (h *AgentServiceHandler) Heartbeat(ctx context.Context, req *pb.HeartbeatRequest) (*pb.HeartbeatResponse, error) {
	log.Printf("[HEARTBEAT] 💓 Nhận heartbeat từ Agent '%s' (%s - %s)",
		req.AgentId, req.Hostname, req.IpAddress)

	now := time.Now()

	// 1. Cập nhật trạng thái Agent trong Redis (với TTL)
	if err := h.redis.SetAgentOnline(ctx, req.AgentId); err != nil {
		log.Printf("[HEARTBEAT] ⚠️ Lỗi cập nhật Redis cho Agent '%s': %v", req.AgentId, err)
	}

	// 2. Lưu thêm heartbeat data vào Redis Hash (CPU, RAM, version)
	heartbeatData := map[string]interface{}{
		"hostname":      req.Hostname,
		"ip_address":    req.IpAddress,
		"os_type":       req.OsType,
		"agent_version": req.AgentVersion,
		"cpu_usage":     req.CpuUsage,
		"memory_usage":  req.MemoryUsage,
		"last_seen":     now.Format(time.RFC3339),
	}
	_ = h.redis.SetAgentHeartbeatData(ctx, req.AgentId, heartbeatData)

	// 3. Cập nhật hoặc tạo mới Agent trong PostgreSQL
	agent := models.Agent{
		ID:              req.AgentId,
		Hostname:        req.Hostname,
		IPAddress:       req.IpAddress,
		OSType:          req.OsType,
		Status:          "online",
		LastHeartbeatAt: &now,
		AgentVersion:    req.AgentVersion,
	}

	// Upsert: Tạo mới nếu chưa tồn tại, cập nhật nếu đã có
	result := h.db.Where("id = ?", req.AgentId).First(&models.Agent{})
	if result.Error == gorm.ErrRecordNotFound {
		// Agent mới → INSERT
		if err := h.db.Create(&agent).Error; err != nil {
			log.Printf("[HEARTBEAT] ❌ Không thể đăng ký Agent '%s': %v", req.AgentId, err)
		} else {
			log.Printf("[HEARTBEAT] 🆕 Agent mới đã đăng ký: '%s' (%s)", req.AgentId, req.Hostname)
		}
	} else {
		// Agent đã tồn tại → UPDATE
		h.db.Model(&models.Agent{}).Where("id = ?", req.AgentId).Updates(map[string]interface{}{
			"hostname":          req.Hostname,
			"ip_address":        req.IpAddress,
			"os_type":           req.OsType,
			"status":            "online",
			"last_heartbeat_at": &now,
			"agent_version":     req.AgentVersion,
		})
	}

	// 4. Trả về response cho Agent
	return &pb.HeartbeatResponse{
		Acknowledged:             true,
		HeartbeatIntervalSeconds: 30, // Agent nên gửi heartbeat mỗi 30 giây
		Message:                  "OK",
	}, nil
}

File: soc-server-python/app/grpc_server/handler.py
# ==============================================================================
# app/grpc_server/handler.py - gRPC Service Handler (AgentServiceServicer)
# Tương đương: internal/agent_grpc/handler.go
#
# Implement AgentService với 2 RPC:
#   1. StreamEvents (Bidirectional Streaming):
#      - Nhận stream Sysmon/Event Log từ Agent
#      - Parse JSON → Gọi RuleEngine.evaluate_log() → Tạo Alert → Broadcast
#      - Lắng nghe command queue → Đẩy CommandResponse xuống Agent
#   2. Heartbeat (Unary):
#      - Cập nhật last_heartbeat_at trong MySQL
#      - Cập nhật agent status = "online" trong Redis (với TTL)
# ==============================================================================

import asyncio
import json
import logging
import uuid
from datetime import datetime
from typing import Optional, Callable, AsyncIterator

import grpc
from sqlalchemy import select

from app.grpc_server import agent_pb2, agent_pb2_grpc
from app.grpc_server.connection_manager import ConnectionManager, ActiveConnection
from app.database import AsyncSessionLocal
from app.models.agent import Agent
from app.models.alert import Alert
from app.models.audit_log import AuditLog
from app.rules.engine import RuleEngine

logger = logging.getLogger(__name__)

# Type alias cho callback khi có alert mới
AlertCallback = Callable[[dict], None]


class AgentServiceHandler(agent_pb2_grpc.AgentServiceServicer):
    """
    Implement AgentServiceServicer interface từ protobuf.
    Tương đương: AgentServiceHandler trong handler.go
    """

    def __init__(
        self,
        rule_engine: RuleEngine,
        conn_manager: ConnectionManager,
        on_new_alert: Optional[AlertCallback] = None,
        redis_client=None,
    ):
        self.rule_engine = rule_engine
        self.conn_manager = conn_manager
        self.on_new_alert = on_new_alert
        self.redis = redis_client

    # ==========================================================================
    # StreamEvents - Bidirectional Streaming RPC
    # Agent ──stream EventRequest──→ Server (receive loop)
    #                                  ↓ evaluate_log()
    #                                  ↓ Match? → Create Alert → Broadcast
    # Agent ←──stream CommandResponse── Server (send loop)
    #                                  ↑ Lắng nghe cmd_queue
    # ==========================================================================
    async def StreamEvents(
        self,
        request_iterator: AsyncIterator[agent_pb2.EventRequest],
        context: grpc.aio.ServicerContext,
    ):
        """Bidirectional streaming: nhận events, gửi commands"""
        conn: Optional[ActiveConnection] = None
        agent_id = ""
        registered = False

        try:
            # ===== Goroutine 2 (Python Task): Sender - Đẩy commands xuống Agent =====
            sender_task: Optional[asyncio.Task] = None

            async def command_sender():
                """Background task lắng nghe cmd_queue và gửi qua stream"""
                try:
                    while not conn.done.is_set():
                        try:
                            cmd_dict = await asyncio.wait_for(
                                conn.cmd_queue.get(), timeout=1.0
                            )
                            # Build protobuf CommandResponse từ dict
                            cmd_pb = agent_pb2.CommandResponse(
                                command_id=cmd_dict.get("command_id", str(uuid.uuid4())),
                                command_type=cmd_dict.get("command_type", 0),
                                target=cmd_dict.get("target", ""),
                                parameters=cmd_dict.get("parameters", {}),
                                timestamp=int(datetime.utcnow().timestamp() * 1000),
                            )
                            await context.write(cmd_pb)
                            logger.info(
                                f"[STREAM] 📤 Đã gửi command '{cmd_pb.command_type}' "
                                f"xuống Agent '{agent_id}'"
                            )
                        except asyncio.TimeoutError:
                            # Timeout bình thường, tiếp tục vòng lặp
                            continue
                        except Exception as e:
                            logger.error(
                                f"[STREAM] ❌ Lỗi gửi command xuống Agent '{agent_id}': {e}"
                            )
                            return
                except asyncio.CancelledError:
                    pass

            # ===== Main Loop: Nhận events từ Agent stream =====
            async for event in request_iterator:
                # Lần nhận đầu tiên: Register Agent connection
                if not registered:
                    agent_id = event.agent_id
                    conn = ActiveConnection(
                        agent_id=agent_id,
                        hostname=event.hostname,
                        ip_address=event.ip_address,
                        context=context,
                    )

                    # Đăng ký connection vào Connection Manager
                    await self.conn_manager.register(conn)
                    registered = True

                    # Bắt đầu sender task
                    sender_task = asyncio.create_task(command_sender())

                    # Cập nhật trạng thái Agent online trong DB và Redis
                    await self._set_agent_online(
                        agent_id=agent_id,
                        hostname=event.hostname,
                        ip_address=event.ip_address,
                    )

                    logger.info(
                        f"[STREAM] 🟢 Agent '{agent_id}' ({event.hostname}) "
                        f"đã bắt đầu stream events"
                    )

                # ===== Xử lý Event: Parse JSON → Evaluate Rule =====
                await self._process_event(event, agent_id)

        except grpc.aio.AbortError:
            pass
        except Exception as e:
            logger.error(f"[STREAM] ❌ Lỗi stream Agent '{agent_id}': {e}")
        finally:
            # Cleanup khi stream đóng
            if registered and conn:
                conn.done.set()
                if sender_task and not sender_task.done():
                    sender_task.cancel()
                await self.conn_manager.unregister(agent_id)
                await self._set_agent_offline(agent_id)
                logger.info(f"[STREAM] 🔴 Agent '{agent_id}' đã ngắt kết nối stream")

    async def _process_event(self, event: agent_pb2.EventRequest, agent_id: str):
        """
        Xử lý 1 event nhận từ Agent.
        Parse raw_payload JSON → Gọi RuleEngine → Tạo Alert nếu match
        Tương đương: processEvent() trong handler.go
        """
        # Parse raw_payload (JSON string) thành dict
        try:
            raw_log = json.loads(event.raw_payload) if event.raw_payload else {}
        except json.JSONDecodeError:
            logger.warning(
                f"[STREAM] ⚠️ Lỗi parse raw_payload JSON từ Agent '{agent_id}'"
            )
            raw_log = {"raw_data": event.raw_payload}

        # Bổ sung metadata vào raw_log để Rule Engine có thêm context
        raw_log["event_type"] = event.event_type
        raw_log["hostname"] = event.hostname
        raw_log["ip_address"] = event.ip_address
        raw_log["agent_id"] = agent_id

        # Metric là telemetry phục vụ biểu đồ — không phải security alert
        if event.event_type == "system_metric":
            return

        # ===== Gọi Rule Engine đánh giá event =====
        alerts_data, matched = self.rule_engine.evaluate_log(raw_log, agent_id)

        if not matched:
            # Event không nguy hiểm vẫn được lưu để analyst theo dõi
            raw_payload_json = json.dumps(raw_log)
            alerts_data = [
                {
                    "id": str(uuid.uuid4()),
                    "agent_id": agent_id,
                    "rule_id": "EVENT-OBSERVED",
                    "event_type": event.event_type or "observed",
                    "severity": "low",
                    "raw_payload": raw_payload_json,
                    "status": "new",
                    "title": f"[LOW] Event observed: {event.event_type}",
                    "description": "Sự kiện được ghi nhận từ Agent nhưng không khớp detection rule nguy hiểm.",
                    "mitre_tactic": "",
                    "mitre_technique_id": "",
                }
            ]

        # ===== Lưu Alerts vào DB và dispatch =====
        async with AsyncSessionLocal() as db:
            for alert_data in alerts_data:
                alert = Alert(
                    id=alert_data.get("id", str(uuid.uuid4())),
                    agent_id=alert_data["agent_id"],
                    rule_id=alert_data.get("rule_id"),
                    event_type=alert_data["event_type"],
                    severity=alert_data["severity"],
                    raw_payload=alert_data.get("raw_payload", ""),
                    status=alert_data.get("status", "new"),
                    title=alert_data.get("title", ""),
                    description=alert_data.get("description", ""),
                    mitre_tactic=alert_data.get("mitre_tactic", ""),
                    mitre_technique_id=alert_data.get("mitre_technique_id", ""),
                    created_at=datetime.utcnow(),
                )
                db.add(alert)

                try:
                    await db.flush()

                    logger.info(
                        f"[STREAM] 🚨 ALERT CREATED: [{alert.severity.upper()}] "
                        f"{alert.title} (Agent: {agent_id}, Rule: {alert.rule_id})"
                    )

                    # ===== Ghi Audit Log =====
                    audit_details = json.dumps({
                        "event_type": alert.event_type,
                        "rule_id": alert.rule_id,
                        "severity": alert.severity,
                    })
                    audit = AuditLog(
                        id=str(uuid.uuid4()),
                        event_type="alert_created",
                        source="rule_based",
                        action="log_only",
                        actor="rule_engine",
                        details=audit_details,
                        alert_id=alert.id,
                        agent_id=agent_id,
                        created_at=datetime.utcnow(),
                    )
                    db.add(audit)

                    # ===== Gọi callback thông báo Alert mới (WebSocket broadcast + SOAR) =====
                    if self.on_new_alert:
                        asyncio.create_task(
                            self._safe_alert_callback(alert_data)
                        )
                except Exception as e:
                    logger.error(f"[STREAM] ❌ Lỗi lưu Alert vào database: {e}")
                    await db.rollback()
                    continue

            try:
                await db.commit()
            except Exception as e:
                logger.error(f"[STREAM] ❌ Lỗi commit batch alerts: {e}")

    async def _safe_alert_callback(self, alert_data: dict):
        """Gọi alert callback an toàn trong background"""
        try:
            if asyncio.iscoroutinefunction(self.on_new_alert):
                await self.on_new_alert(alert_data)
            else:
                self.on_new_alert(alert_data)
        except Exception as e:
            logger.error(f"[STREAM] ❌ Lỗi alert callback: {e}")

    # ==========================================================================
    # Heartbeat - Unary RPC
    # Agent gửi heartbeat định kỳ (mỗi 30 giây) để báo cáo trạng thái
    # Tương đương: Heartbeat() trong handler.go
    # ==========================================================================
    async def Heartbeat(
        self,
        request: agent_pb2.HeartbeatRequest,
        context: grpc.aio.ServicerContext,
    ) -> agent_pb2.HeartbeatResponse:
        """Unary RPC: nhận heartbeat từ Agent, cập nhật DB và Redis"""
        now = datetime.utcnow()

        # ===== 1. Cập nhật Redis nếu có =====
        if self.redis:
            try:
                await self.redis.set(
                    f"agent:{request.agent_id}:online",
                    "1",
                    ex=90,  # TTL 90 giây (3× heartbeat interval)
                )
                heartbeat_data = {
                    "hostname": request.hostname,
                    "ip_address": request.ip_address,
                    "os_type": request.os_type,
                    "agent_version": request.agent_version,
                    "cpu_usage": str(request.cpu_usage),
                    "memory_usage": str(request.memory_usage),
                    "last_seen": now.isoformat(),
                }
                await self.redis.hset(
                    f"agent:{request.agent_id}:heartbeat", mapping=heartbeat_data
                )
                await self.redis.expire(f"agent:{request.agent_id}:heartbeat", 90)
            except Exception as e:
                logger.warning(
                    f"[HEARTBEAT] ⚠️ Lỗi cập nhật Redis cho Agent '{request.agent_id}': {e}"
                )

        # ===== 2. Upsert Agent trong MySQL =====
        async with AsyncSessionLocal() as db:
            try:
                stmt = select(Agent).where(Agent.id == request.agent_id)
                res = await db.execute(stmt)
                existing = res.scalar_one_or_none()

                if existing is None:
                    # Agent mới → INSERT
                    agent = Agent(
                        id=request.agent_id,
                        hostname=request.hostname,
                        ip_address=request.ip_address,
                        os_type=request.os_type,
                        status="online",
                        last_heartbeat_at=now,
                        agent_version=request.agent_version,
                    )
                    db.add(agent)
                    logger.info(
                        f"[HEARTBEAT] 🆕 Agent mới đã đăng ký: '{request.agent_id}' "
                        f"({request.hostname})"
                    )
                else:
                    # Agent đã tồn tại → UPDATE
                    existing.hostname = request.hostname
                    existing.ip_address = request.ip_address
                    existing.os_type = request.os_type
                    existing.status = "online"
                    existing.last_heartbeat_at = now
                    existing.agent_version = request.agent_version

                await db.commit()
            except Exception as e:
                logger.error(
                    f"[HEARTBEAT] ❌ Không thể lưu Agent '{request.agent_id}': {e}"
                )
                await db.rollback()

        # ===== 3. Trả về response cho Agent =====
        return agent_pb2.HeartbeatResponse(
            acknowledged=True,
            heartbeat_interval_seconds=30,
            message="OK",
        )

    # ==========================================================================
    # Helper: Cập nhật Agent status trong DB
    # ==========================================================================
    async def _set_agent_online(self, agent_id: str, hostname: str, ip_address: str):
        """Cập nhật Agent thành online khi bắt đầu stream"""
        async with AsyncSessionLocal() as db:
            try:
                stmt = select(Agent).where(Agent.id == agent_id)
                res = await db.execute(stmt)
                agent = res.scalar_one_or_none()
                if agent:
                    agent.status = "online"
                    agent.ip_address = ip_address
                    agent.hostname = hostname
                else:
                    db.add(Agent(
                        id=agent_id,
                        hostname=hostname,
                        ip_address=ip_address,
                        os_type="unknown",
                        status="online",
                    ))
                await db.commit()
            except Exception as e:
                logger.error(f"[STREAM] ❌ Lỗi cập nhật Agent online: {e}")

        # Cập nhật Redis
        if self.redis:
            try:
                await self.redis.set(f"agent:{agent_id}:online", "1", ex=90)
            except Exception:
                pass

    async def _set_agent_offline(self, agent_id: str):
        """Cập nhật Agent thành offline khi stream đóng"""
        async with AsyncSessionLocal() as db:
            try:
                stmt = select(Agent).where(Agent.id == agent_id)
                res = await db.execute(stmt)
                agent = res.scalar_one_or_none()
                if agent:
                    agent.status = "offline"
                    await db.commit()
            except Exception as e:
                logger.error(f"[STREAM] ❌ Lỗi cập nhật Agent offline: {e}")

        # Xóa key Redis
        if self.redis:
            try:
                await self.redis.delete(f"agent:{agent_id}:online")
            except Exception:
                pass
